src/hotspot/share/gc/serial/defNewGeneration.cpp

```diff
@@ -140,12 +140,16 @@ class CLDScanClosure: public CLDClosure {
 };
 
 class IsAliveClosure: public BoolObjectClosure {
-  HeapWord*         _young_gen_end;
+  DefNewGeneration* _young_gen;
 public:
-  IsAliveClosure(DefNewGeneration* g): _young_gen_end(g->reserved().end()) {}
+  IsAliveClosure(DefNewGeneration* g):
+    _young_gen(g) {}
 
   bool do_object_b(oop p) {
-    return cast_from_oop<HeapWord*>(p) >= _young_gen_end || p->is_forwarded();
+    HeapWord* heap_word_ptr = cast_from_oop<HeapWord*>(p);
+    bool is_in_young_gen = _young_gen->is_in_reserved((void*)heap_word_ptr);
```

Comment:
The check for only the boundary is intentional, and guided by performance. One comparison/memory load in the original code is faster than the two memory loads/comparisons (both bounds of the reserved area) plus the eventual indirect load via `_young_gen` is just slower.

The same reason why most of the places it is used store a local copy of the generation boundary.

This adds up for hundreds of thousands of checks/references to evacuate; at least it did last time years ago.

If the goal is just factoring out the check to make it change for all locations whenever heap layout is changed, I would prefer adding a helper method in `SerialHeap` for example that's easily inlinable for the compiler.

Reply:
> If the goal is just factoring out the check to make it change for all locations whenever heap layout is changed, I would prefer adding a helper method in `SerialHeap` for example that's easily inlinable for the compiler.

Thanks for the explanation. I'll look into a different approach like an inlinable helper method.